File: src/main/java/com/technews/aggregate/releases/service/ReleasesService.java
package com.technews.aggregate.releases.service;

import com.technews.aggregate.releases.domain.Release;
import com.technews.aggregate.releases.domain.repository.ReleasesRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.CollectionUtils;

import java.util.List;

@Slf4j
@Service
@RequiredArgsConstructor
public class ReleasesService {

    private final ReleasesRepository releasesRepository;

    @Transactional(readOnly = true)
    public Page<Release> findAllRelease(
            final int page, final int size, final List<String> categories) {
        final PageRequest pageable = PageRequest.of(
                page, size,
                Sort.by("createdDt").descending().and(Sort.by("version").descending()));

        return getReleasePage(categories, pageable);
    }

    private Page<Release> getReleasePage(final List<String> categories, final PageRequest pageable) {
        if (CollectionUtils.isEmpty(categories)) {
            return releasesRepository.findAll(pageable);
        }

        return releasesRepository.findByProjectIn(categories, pageable);
    }
}

File: src/main/kotlin/com/technews/aggregate/releases/service/ReleasesService.kt
package com.technews.aggregate.releases.service

import com.technews.aggregate.releases.domain.repository.ReleasesRepository
import com.technews.aggregate.releases.dto.ReleaseResponse
import mu.KotlinLogging
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

private val logger = KotlinLogging.logger { }

@Service
class ReleasesService(
    private val releasesRepository: ReleasesRepository,
) {
    @Transactional(readOnly = true)
    fun findAllRelease(
        pageable: PageRequest,
        categories: List<String>? = emptyList(),
    ): Page<ReleaseResponse> {
        val releasePage = if (categories.isNullOrEmpty()) {
            releasesRepository.findAll(pageable)
        } else {
            releasesRepository.findByProjectIn(categories, pageable)
        }

        return releasePage.map { release ->
            ReleaseResponse(
                id = release.id ?: "",
                project = release.project,
                version = release.version,
                date = release.date,
                url = release.url,
                tags = release.tags,
                shared = release.shared,
                createdDt = release.createdDt,
            )
        }
    }
}
